azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -992,60 +1015,52 @@ def _send_new_connect_op_down(self):
         def on_connect_complete(op, error):
             this = self_weakref()
             if this:
+                logger.debug(
+                    "{}({}): on_connect_complete error={} state={} never_conencted={} connected={} ".format(
+                        this.name,
+                        op.name,
+                        error,
+                        this.state,
+                        this.never_connected,
+                        this.pipeline_root.connected,
+                    )
+                )
                 if error:
-                    if this.state == ReconnectState.NEVER_CONNECTED:
-                        logger.info(
-                            "{}({}): error on first connection.  Not triggering reconnection".format(
-                                this.name, op.name
-                            )
-                        )
+                    if this.never_connected:
+                        # any error on a first connection is fatal
+                        this.state = ReconnectState.LOGICALLY_DISCONNECTED
+                        this._clear_reconnect_timer()
                         this._complete_waiting_connect_ops(error)
                     elif type(error) in transient_connect_errors:
-                        logger.info(
-                            "{}({}): State is {}.  Connect failed with transient error. Triggering reconnect timer".format(
-                                self.name, op.name, self.state
-                            )
-                        )
+                        # transient errors cause a reconnect attempt
                         self.state = ReconnectState.WAITING_TO_RECONNECT
-                        self._start_reconnect_timer()
-
-                    elif this.state == ReconnectState.WAITING_TO_RECONNECT:
-                        logger.info(
-                            "{}({}): non-tranient error.  Failing all waiting ops.n".format(
-                                this.name, op.name
-                            )
-                        )
-                        self.state = ReconnectState.CONNECTED_OR_DISCONNECTED
-                        self._clear_reconnect_timer()
-                        this._complete_waiting_connect_ops(error)
-
+                        self._start_reconnect_timer(self.delayed_reconnect_delay)
                     else:
-                        logger.info(
-                            "{}({}): State is {}.  Connection failed. Not triggering reconnection".format(
-                                this.name, op.name, this.state
-                            )
-                        )
+                        # all others are fatal
+                        this.state = ReconnectState.LOGICALLY_DISCONNECTED
+                        this._clear_reconnect_timer()
                         this._complete_waiting_connect_ops(error)
```

Comment:
>this._complete_waiting_connect_ops(error) [](start = 24, length = 41)

This seems like it would include, as mentioned in a comment above, OperationCancelled operations, which represent a disconnect somehow.

So, reaching a disconnect state by connection failure causes the disconnect operation to fail?

Reply:
Related to my previous response.  The DisconnectOperation is never set as a waiting_connectoin_op.  The disconnect causes waiting ops to be cancelled, but it's never a waiting op itself.

---
In reply to: [439687473](https://github.com/Azure/azure-iot-sdk-python/pull/576#discussion_r439687473) [](ancestors = 439687473)